[WebAuthn] Add support for authenticators over CCID

pascoej opened the pull request:
#### c0f5fc64aac387a18a1ec9f54e013847964b4067
<pre>
[WebAuthn] Add support for authenticators over CCID
<a href="https://bugs.webkit.org/show_bug.cgi?id=242365">https://bugs.webkit.org/show_bug.cgi?id=242365</a>
rdar://82529212

Reviewed by Brent Fulgham.

This patch adds support for authenticators over CCID. This allows use of credentials
on smart cards. The transport used is determined by how the card responds to the
&quot;Get Data Command,&quot; which returns a uid with contactless cards.

Added layout tests and tested manually with a smart card (contact and contactless), as well
as a Yubikey 5c over NFC.

* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-ccid.https-expected.txt: Added.
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-ccid.https.html: Added.
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-ccid.https-expected.txt: Added.
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-ccid.https.html: Added.
* LayoutTests/http/wpt/webauthn/resources/util.js:
* Source/WebCore/Modules/webauthn/AuthenticatorTransport.h:
* Source/WebCore/Modules/webauthn/AuthenticatorTransport.idl:
* Source/WebCore/Modules/webauthn/WebAuthenticationConstants.h:
* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.cpp:
(fido::toString):
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::convertStringToAuthenticatorTransport):
* Source/WebCore/testing/MockWebAuthenticationConfiguration.h:
(WebCore::MockWebAuthenticationConfiguration::CcidConfiguration::encode const):
(WebCore::MockWebAuthenticationConfiguration::CcidConfiguration::decode):
(WebCore::MockWebAuthenticationConfiguration::encode const):
(WebCore::MockWebAuthenticationConfiguration::decode):
* Source/WebCore/testing/MockWebAuthenticationConfiguration.idl:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::WebCore::collectTransports):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorTransportService.cpp:
(WebKit::AuthenticatorTransportService::create):
(WebKit::AuthenticatorTransportService::createMock):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidConnection.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.h.
(WebKit::CcidConnection::contactless const):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidConnection.mm: Added.
(WebKit::fido::compareVersion):
(WebKit::CcidConnection::create):
(WebKit::CcidConnection::CcidConnection):
(WebKit::CcidConnection::~CcidConnection):
(WebKit::CcidConnection::detectContactless):
(WebKit::CcidConnection::trySelectFidoApplet):
(WebKit::CcidConnection::transact const):
(WebKit::CcidConnection::stop const):
(WebKit::CcidConnection::restartPolling):
(WebKit::CcidConnection::startPolling):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidService.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.h.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidService.mm: Added.
(WebKit::CcidService::CcidService):
(WebKit::CcidService::~CcidService):
(WebKit::CcidService::didConnectTag):
(WebKit::CcidService::startDiscoveryInternal):
(WebKit::CcidService::restartDiscoveryInternal):
(WebKit::CcidService::platformStartDiscovery):
(WebKit::CcidService::onValidCard):
(WebKit::CcidService::updateSlots):
(-[_WKSmartCardSlotObserver initWithService:]):
(-[_WKSmartCardSlotObserver observeValueForKeyPath:ofObject:change:context:]):
(-[_WKSmartCardSlotStateObserver initWithService:slot:]):
(-[_WKSmartCardSlotStateObserver observeValueForKeyPath:ofObject:change:context:]):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toASCDescriptor):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockAuthenticatorManager.cpp:
(WebKit::MockAuthenticatorManager::filterTransports const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockCcidService.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.idl.
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockCcidService.mm: Added.
(-[_WKMockTKSmartCard initWithService:]):
(-[_WKMockTKSmartCard beginSessionWithReply:]):
(-[_WKMockTKSmartCard transmitRequest:reply:]):
(WebKit::MockCcidService::MockCcidService):
(WebKit::MockCcidService::platformStartDiscovery):
(WebKit::MockCcidService::nextReply):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapCcidDriver.cpp: Added.
(WebKit::CtapCcidDriver::CtapCcidDriver):
(WebKit::CtapCcidDriver::transact):
(WebKit::CtapCcidDriver::respondAsync const):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapCcidDriver.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.h.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/252425@main">https://commits.webkit.org/252425@main</a>
</pre>

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidConnection.mm:
```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
```
(C) 2022 (here and elsewhere).

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidService.mm:
```diff
@@ -0,0 +1,182 @@
+
+- (void)observeValueForKeyPath:(id)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+    UNUSED_PARAM(object);
```
It seems like 'object' is used, probably don't need UNUSED_PARAM for it.

brentfulgham reviewed (approved):
Looks good. A couple of nits to be adjusted before you land.

pascoej committed:
[WebAuthn] Add support for authenticators over CCID
https://bugs.webkit.org/show_bug.cgi?id=242365
rdar://82529212

Reviewed by Brent Fulgham.

This patch adds support for authenticators over CCID. This allows use of credentials
on smart cards. The transport used is determined by how the card responds to the
"Get Data Command," which returns a uid with contactless cards.

Added layout tests and tested manually with a smart card (contact and contactless), as well
as a Yubikey 5c over NFC.

* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-ccid.https-expected.txt: Added.
* LayoutTests/http/wpt/webauthn/public-key-credential-create-success-ccid.https.html: Added.
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-ccid.https-expected.txt: Added.
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-ccid.https.html: Added.
* LayoutTests/http/wpt/webauthn/resources/util.js:
* Source/WebCore/Modules/webauthn/AuthenticatorTransport.h:
* Source/WebCore/Modules/webauthn/AuthenticatorTransport.idl:
* Source/WebCore/Modules/webauthn/WebAuthenticationConstants.h:
* Source/WebCore/Modules/webauthn/fido/AuthenticatorGetInfoResponse.cpp:
(fido::toString):
* Source/WebCore/Modules/webauthn/fido/DeviceResponseConverter.cpp:
(fido::convertStringToAuthenticatorTransport):
* Source/WebCore/testing/MockWebAuthenticationConfiguration.h:
(WebCore::MockWebAuthenticationConfiguration::CcidConfiguration::encode const):
(WebCore::MockWebAuthenticationConfiguration::CcidConfiguration::decode):
(WebCore::MockWebAuthenticationConfiguration::encode const):
(WebCore::MockWebAuthenticationConfiguration::decode):
* Source/WebCore/testing/MockWebAuthenticationConfiguration.idl:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::WebCore::collectTransports):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorTransportService.cpp:
(WebKit::AuthenticatorTransportService::create):
(WebKit::AuthenticatorTransportService::createMock):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidConnection.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.h.
(WebKit::CcidConnection::contactless const):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidConnection.mm: Added.
(WebKit::fido::compareVersion):
(WebKit::CcidConnection::create):
(WebKit::CcidConnection::CcidConnection):
(WebKit::CcidConnection::~CcidConnection):
(WebKit::CcidConnection::detectContactless):
(WebKit::CcidConnection::trySelectFidoApplet):
(WebKit::CcidConnection::transact const):
(WebKit::CcidConnection::stop const):
(WebKit::CcidConnection::restartPolling):
(WebKit::CcidConnection::startPolling):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidService.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.h.
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidService.mm: Added.
(WebKit::CcidService::CcidService):
(WebKit::CcidService::~CcidService):
(WebKit::CcidService::didConnectTag):
(WebKit::CcidService::startDiscoveryInternal):
(WebKit::CcidService::restartDiscoveryInternal):
(WebKit::CcidService::platformStartDiscovery):
(WebKit::CcidService::onValidCard):
(WebKit::CcidService::updateSlots):
(-[_WKSmartCardSlotObserver initWithService:]):
(-[_WKSmartCardSlotObserver observeValueForKeyPath:ofObject:change:context:]):
(-[_WKSmartCardSlotStateObserver initWithService:slot:]):
(-[_WKSmartCardSlotStateObserver observeValueForKeyPath:ofObject:change:context:]):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toASCDescriptor):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockAuthenticatorManager.cpp:
(WebKit::MockAuthenticatorManager::filterTransports const):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockCcidService.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.idl.
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockCcidService.mm: Added.
(-[_WKMockTKSmartCard initWithService:]):
(-[_WKMockTKSmartCard beginSessionWithReply:]):
(-[_WKMockTKSmartCard transmitRequest:reply:]):
(WebKit::MockCcidService::MockCcidService):
(WebKit::MockCcidService::platformStartDiscovery):
(WebKit::MockCcidService::nextReply):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapCcidDriver.cpp: Added.
(WebKit::CtapCcidDriver::CtapCcidDriver):
(WebKit::CtapCcidDriver::transact):
(WebKit::CtapCcidDriver::respondAsync const):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapCcidDriver.h: Copied from Source/WebCore/Modules/webauthn/AuthenticatorTransport.h.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/252425@main
